bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java

```diff
@@ -0,0 +1,240 @@
+     * @throws ParseException - when the expression can not be parsed correctly
+     */
+    public AbstractExpression(String expression, String delimiters, Date startDate, TimeZone timeZone,
+            int minimumCandidates) throws ParseException {
+        super();
+
+        if (expression == null) {
+            throw new IllegalArgumentException("The expression cannot be null");
+        }
+
+        this.expression = expression;
+        this.delimiters = delimiters;
+        this.startDate = startDate;
+        this.timeZone = timeZone;
+        this.minimumCandidates = minimumCandidates;
+
+        setStartDate(startDate);
+        setTimeZone(timeZone);
+        parseExpression(expression);
+    }
+
+    @Override
+    public final Date getStartDate() {
+        if (startDate == null) {
+            startDate = Calendar.getInstance(getTimeZone()).getTime();
+        }
+        return startDate;
+    }
+
+    @Override
+    public void setStartDate(Date startDate) {
+        if (startDate == null) {
+            throw new IllegalArgumentException("The start date of the rule can not be null");
+        }
+        this.startDate = startDate;
+    }
+
+    @Override
+    public TimeZone getTimeZone() {
+        if (timeZone == null) {
+            timeZone = TimeZone.getDefault();
+        }
+
+        return timeZone;
+    }
+
+    @Override
+    public void setTimeZone(TimeZone timeZone) {
+        this.timeZone = timeZone;
+    }
+
+    @Override
+    public String getExpression() {
+        return expression;
+    }
+
+    @Override
+    public void setExpression(String expression) throws ParseException {
+        this.expression = expression;
+        parseExpression(expression);
+    }
+
+    @Override
+    public String toString() {
+        return expression;
+    }
+
+    /**
+     * Parse the given expression
+     * 
+     * @param expression - the expression to parse
+     * @throws ParseException - when the expression can not be successfully be parsed
+     * @throws IllegalArgumentException - when expression parts conflict with each other
+     */
+    @SuppressWarnings("unchecked")
+    public void parseExpression(String expression) throws ParseException, IllegalArgumentException {
+
+        StringTokenizer expressionTokenizer = new StringTokenizer(expression, delimiters, false);
+        int position = 0;
+
+        while (expressionTokenizer.hasMoreTokens()) {
+            String token = expressionTokenizer.nextToken().trim();
+            position++;
+            expressionParts.add(parseToken(token, position));
+        }
+
+        Collections.sort(expressionParts);
+
+        validateExpression();
+
+        if (startDate == null) {
+            setStartDate(Calendar.getInstance().getTime());
+        }
+
+        applyExpressionParts();
+
+        continueSearch = true;
+        while (candidates.size() < minimumCandidates && continueSearch) {
```

Comment:
Your are using a protected field for ```continueSearch``` here. As it's an instance variable and there are no synchronized blocks/statements, this will break when getting executed from different threads